### PR TITLE
[system testing] Add provision for system testing negative / false-positive scenarios in packages

### LINKED ISSUE #1308: Reduce false negatives with assert.hit_count
When the `assert.hit_count` condition is satisfied the test completes. But if more data is ingested after the condition is satisfied then the test will not catch it leading to a false negative result.

To avoid this problem the test case runner could perform a one or more additional hit count checks after some delay to help ensure that the test fails when more data is ingested than is expected.

### PR DESCRIPTION
**Pull request by bhapas**
This PR changes the way the `system testing` asserts at the number of hits in ElasticSearch when a system test is run.

Earlier, The testrunner exited when `enough` docs are found as per the config and there were more docs that got ingested after `assertHitCount` condition satisfied.

Now, The testrunner tries to get all the docs before checking for the `assertHitCount` and fails if the docs are more/less than the configured count

The `retryTicker` in the `waitUntilTrue` function is changed to `5 seconds` meaning , the retry interval for each poll to ES has been  changed to `5 seconds` and so if for `10 seconds`, there is no change in number of docs , the test passes.

### How to test

Run `PACKAGE_TEST_TYPE=false_positives PACKAGE_UNDER_TEST=httpjson_false_positive_asserts ./scripts/test-check-packages-false-positives.sh`

The `test-check-packages.sh` file has been modified to catch a failure from the new package `httpjson_false_positive_asserts` and a `testcase failure` is considered as `success scenario`

- Fixes #1308

### CODE REVIEW

**Comment by jsoriano** [comment]
test integrations

File: scripts/test-check-false-positives.sh
Comment (jsoriano) [inline scripts/test-check-false-positives.sh:23]: Nice to have a test for false positive cases, thanks! Not needed to change it here by now, but would it be possible to reuse this script to be able to add more test packages?

File: internal/testrunner/runners/system/runner.go
Comment (jsoriano) [inline internal/testrunner/runners/system/runner.go:990]: `waitUntilTrue` is used in more places and this change will affect them, could we maybe add the additional wait in the function to check the hits?

File: internal/testrunner/runners/system/runner.go
Comment (jsoriano) [inline internal/testrunner/runners/system/runner.go:635]: This is going to make all tests to wait till timeout or till the count of documents is stable, isn't it? Could we continue taking into acount if `config.Assert.HitCount` is configured so we don't extend the wait for the rest of cases?

Maybe to solve this issue it is enough with adding a sleep when `hits.size() >= config.Assert.HitCount`?

**Review by jsoriano — CHANGES_REQUESTED** [review]
Nice to have the first negative test :slightly_smiling_face:

I think these changes are going to affect the timings of all tests, and not only the ones with hit count assertions, could we change it so these changes only affect tests with hit count assertions?

**Comment by mrodm** [comment]
test integrations

File: scripts/test-check-false-positives.sh
Comment (bhapas) [reply scripts/test-check-false-positives.sh:23]: Yes. It should be possible to add more packages under `PACKAGE_TEST_TYPE` `false_positives`

File: scripts/test-check-false-positives.sh
Comment (mrodm) [inline scripts/test-check-false-positives.sh:59]: In `pipeline.trigger.integration.tests.sh` script, PACKAGE_UNDER_TEST is not being set for the loop where this package has been added.

If it is needed, it must be added the `-p` parameter here in this line

https://github.com/elastic/elastic-package/pull/1333/files#diff-4e2e09cf3f17805ad30f828b15ea267ba7a186ff3f5129b9ecba76ef7317718bR41

Currently, when PACKAGE_UNDER_TEST is being set is because it is required/needed to just execute one package per stage in Jenkins, as it is done for the test packages under `test/packages/parallel`. These packages have special requirements (e.g. logs are updated to a special artifact buckets).

If it is needed to execute every package under `false_positives` in a different stage, it could be added a new loop there in that script for this false_positives packages. Maybe something like:

```bash
pushd test/packages/false_positives > /dev/null
for package in $(find . -maxdepth 1 -mindepth 1 -type d) ; do
    package_name=$(basename ${package})
    echo "      - label: \":go: Running integration test: ${package_name}\""
    echo "        key: \"integration-false_positives-${package_name}\""
    echo "        command: ./.buildkite/scripts/integration_tests.sh -t test-check-packages-false-positives -p ${package_name}"
    echo "        agents:"
    echo "          provider: \"gcp\""
    echo "        artifact_paths:"
    echo "          - build/test-results/*.xml"
    echo "          - build/elastic-stack-dump/check-*/logs/*.log"
    echo "          - build/elastic-stack-dump/check-*/logs/fleet-server-internal/**/*"
done
popd > /dev/null
```

File: internal/testrunner/runners/system/runner.go
Comment (bhapas) [reply internal/testrunner/runners/system/runner.go:635]: Breaking this down into scenarios :

### HitCount configured
- Number of docs less than the count in `n` iterations - Timeout happens as the `true` condition is never satisfied - Test fails
- Number of docs equal in 2 successive iterations with a gap of 5 seconds - No more  docs to ingest and then verify if the count is matching - Catching false positives
- Number of docs never equal - Some issue in the test case causing the docs ingestion to go into infinite loop - Timeout and Test fails

### HitCount Not configured
- Docs ingestion complete - Meaning the `hits.size()` is same in 2 consecutive iterations with 1 second each - Adds up 1 second on each test case but provides us with data if we were to validate something else.
- Docs ingestion incomplete - Just picking up those docs which are ingested in the first iteration and validating against them

I think it should not add up a lot of time for each run , as I see the test suites are running in parallel , but we shall have the data ingested for any other validation - Just to avoid testing against incomplete data - WDYT?

File: internal/testrunner/runners/system/runner.go
Comment (bhapas) [reply internal/testrunner/runners/system/runner.go:990]: Will send in a `retryTimeout` from the calling function

Comment (bhapas) [reply internal/testrunner/runners/system/runner.go:990]: Or Maybe a `time.sleep(4 * time.Second)` if config for `assert.hitCount` exists

File: scripts/test-check-false-positives.sh
Comment (bhapas) [reply scripts/test-check-false-positives.sh:59]: These `false_positive` tests should just run as `System tests` for any other package, just that the outcome is inversed.

I have add the `PACKAGE_UNDER_TEST` in https://github.com/elastic/elastic-package/pull/1333/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R81 

and it seems to run as a separate suite atleast. Not sure if it would be necessary to run each package separately? Do you think the the testrunner exits on execution failure / the script failure? Like `exit 1 from test execution`  or `exit 1 from script execution` 


Or Did I misunderstand ?

Screenshot 2023-07-03 at 11 48 14

Comment (bhapas) [reply scripts/test-check-false-positives.sh:59]: I think I got your point now.. 

But as it is now , I don't see a need to execute each package in a separate buildKite stage. As they are just another system tests.. And I think the value of `PACKAGE_UNDER_TEST` is correctly taken in each execution as per https://buildkite.com/elastic/elastic-package/builds/1099#01890bc4-ce48-4b8b-9638-8ae44f58e102/131-2282

**Comment by bhapas** [comment]
test integrations

File: internal/testrunner/runners/system/runner.go
Comment (jsoriano) [inline internal/testrunner/runners/system/runner.go:643]: In this case, should we wait for the hits size to be greater than `config.Assert.HitCount`? Otherwise, if the time between the first and last event takes more than 5 secons, it won't wait for the last hits.
```suggestion
			if hits.size() < config.Assert.HitCount {
				return false, err
			}		
			time.Sleep(4 * time.Second)
			return true, err
```

Comment (bhapas) [reply internal/testrunner/runners/system/runner.go:643]: In this case if the 2 docs are ingested in a span of 5 seconds in 2 iterations , and `assert.hit_count: 2` is set in the config , this code returns true in the first round after it matches `2` , but total docs ingested are 4.

Comment (bhapas) [reply internal/testrunner/runners/system/runner.go:643]: May be can be mitigated this way - WDYT?

```diff
diff --git a/internal/testrunner/runners/system/runner.go b/internal/testrunner/runners/system/runner.go
index cc47cc9..a77af87 100644
--- a/internal/testrunner/runners/system/runner.go
+++ b/internal/testrunner/runners/system/runner.go
@@ -634,6 +634,10 @@ func (r *runner) runTest(config *testConfig, ctxt servicedeployer.ServiceContext
                       if config.Assert.HitCount > 0 {
			if hits.size() == 0 {
				return false, err
			}
 
+                       if hits.size() < config.Assert.HitCount {
+                               return false, err
+                       }
+
                       ret := hits.size() == oldHits
			if !ret {
				oldHits = hits.size()
				time.Sleep(4 * time.Second)
			}

			return ret, err
```

Comment (jsoriano) [reply internal/testrunner/runners/system/runner.go]: The problem I see is that we are not waiting for the number of events to be at least the one we expect. So it may happen that the wait finishes, the count is not the one we expect yet, because we haven't waited enough, and then the later assertion fails, when it should have passed.

This implementation is waiting to have at least one event, and that the count of events is stable during at least 5 seconds. But I think it should also wait till the count of events is at least `assert.hit_count`, or timeout if it never is.

Imagine an scenario with `assert.hit_count: 2`. One event arrives during the first second, second one arrives in ten seconds.
* One event is ingested.
* One iteration happens, oldHits is zero, current hits is one, we set oldHits to one, wait 4 seconds and retry.
* Second iteration happens, oldHits is one, current hits is still one, we return true, the wait finishes.
* Assertion fails because there is only one hit. But it shouldn't because another one is going to arrive in ten seconds, way before the timeout.

I think that this wait should never return true if `config.Assert.HitCount` is set and the current count of hints is below this expected count.

Comment (bhapas) [reply internal/testrunner/runners/system/runner.go]: > The problem I see is that we are not waiting for the number of events to be at least the one we expect. So it may happen that the wait finishes, the count is not the one we expect yet, because we haven't waited enough, and then the later assertion fails, when it should have passed.

Resolved this in the latest comment - https://github.com/elastic/elastic-package/pull/1333#discussion_r1250752641

File: scripts/test-check-false-positives.sh
Comment (bhapas) [reply scripts/test-check-false-positives.sh:59]: Added another CI step

File: internal/testrunner/runners/system/runner.go
Comment (jsoriano) [reply internal/testrunner/runners/system/runner.go]: Ah yes, sorry this would work :slightly_smiling_face: I was writing my comment at the same time and didn't see it.

Comment (bhapas) [reply internal/testrunner/runners/system/runner.go]: Pushed the changes now :-)

File: internal/testrunner/runners/system/runner.go
Comment (jsoriano) [inline internal/testrunner/runners/system/runner.go:639]: Nit. First if is redundant.
```suggestion
			if hits.size() < config.Assert.HitCount {
				return false, err
			}
```

File: .buildkite/pipeline.trigger.integration.tests.sh
Comment (mrodm) [inline .buildkite/pipeline.trigger.integration.tests.sh:63]: Should we do as in `parallel` packages and upload logs to the internal bucket ? cc @jsoriano

File: .buildkite/pipeline.trigger.integration.tests.sh
Comment (mrodm) [inline .buildkite/pipeline.trigger.integration.tests.sh:55]: Should it be indicated here that is from `false positive`? 

```suggestion
    echo "      - label: \":go: Running integration test (false positive): ${package_name}\""
```

File: .buildkite/pipeline.trigger.integration.tests.sh
Comment (jsoriano) [reply .buildkite/pipeline.trigger.integration.tests.sh:63]: Sorry, I missed this comment.

I would do the same as in `parallel`, so we don't forget to do it here if we want to avoid risks. I guess that in `parallel` we are doing this to avoid leaking secrets related to cloud integrations or the terraform test runner.

Comment (bhapas) [reply .buildkite/pipeline.trigger.integration.tests.sh:63]: @jsoriano Does this add any value? I know why we did it in `parallel` , as we used real cloud instances and secrets. These are mere mocked tests , or?

Comment (jsoriano) [reply .buildkite/pipeline.trigger.integration.tests.sh:63]: Tests packages including terraform scenarios are actually executed.

Comment (bhapas) [reply .buildkite/pipeline.trigger.integration.tests.sh:65]: Done.. Pushed

**Comment by bhapas** [comment]
test integrations

File: scripts/test-check-false-positives.sh
Comment (jsoriano) [inline scripts/test-check-false-positives.sh:9]: Should this have been changed after this comment? https://github.com/elastic/elastic-package/pull/1333/files#r1250740254
```suggestion
  elastic-package stack dump -v --output "build/elastic-stack-dump/check-${PACKAGE_UNDER_TEST:-${PACKAGE_TEST_TYPE:-*}}"
```

**Review by jsoriano — COMMENTED** [review]
LGTM, let's wait for a final review by Mario to the CI scripts :innocent: 

Thanks!

File: scripts/test-check-false-positives.sh
Comment (bhapas) [reply scripts/test-check-false-positives.sh:9]: Instead each package is run in a separate small step in parallel

Comment (bhapas) [reply scripts/test-check-false-positives.sh:9]: @mrodm Is this the same as we talked or?

File: .buildkite/pipeline.trigger.integration.tests.sh
Comment (mrodm) [inline .buildkite/pipeline.trigger.integration.tests.sh:65]: According to this thread https://github.com/elastic/elastic-package/pull/1333/files#r1250786653, let's just remove these two from `artifact_paths` and just keep the result xml files, as it happens for `parallel` test packages:
 ```suggestion
```

File: scripts/test-check-false-positives.sh
Comment (mrodm) [reply scripts/test-check-false-positives.sh:9]: @jsoriano is right, in this script the default value for `PACKAGE_UNDER_TEST` should be `*` instead of the `httpjson_false_positive_asserts` package. In that regard, this script would be similar to `scripts/test-check-packages.sh`.

File: scripts/test-check-false-positives.sh
Comment (mrodm) [inline scripts/test-check-false-positives.sh:15]:
```suggestion
  for d in test/packages/${PACKAGE_TEST_TYPE:-false_positives}/${PACKAGE_UNDER_TEST:-*}/; do
```

File: scripts/test-check-false-positives.sh
Comment (mrodm) [inline scripts/test-check-false-positives.sh:40]:
```suggestion
for d in test/packages/${PACKAGE_TEST_TYPE:-false_positives}/${PACKAGE_UNDER_TEST:-*}/; do
```

File: scripts/test-check-false-positives.sh
Comment (mrodm) [inline scripts/test-check-false-positives.sh:59]:
```suggestion
for d in test/packages/${PACKAGE_TEST_TYPE:-false_positives}/${PACKAGE_UNDER_TEST:-*}/; do
```

File: scripts/test-check-false-positives.sh
Comment (mrodm) [inline scripts/test-check-false-positives.sh:23]: Would it be enough to just check for `PACKAGE_TESAT_TYPE` being `false_positives` ? 

That would mean that every test package located under `false_positives`folder would follow these  requirements (so that it needs to fail to be considered succeeded).

```suggestion
  if [ "${PACKAGE_TEST_TYPE:-false_positives}" == "false_positives" ]; then
```

File: test/packages/false_positives/httpjson_false_positive_asserts/_dev/build/build.yml
Comment (mrodm) [inline test/packages/false_positives/httpjson_false_positive_asserts/_dev/build/build.yml:3]: Let's just use tag here. 
There were some issues related to use branches instead of tags

```suggestion
    reference: git@v8.8.0
```

Probably, it results in some failures in the README

File: scripts/test-check-false-positives.sh
Comment (bhapas) [reply scripts/test-check-false-positives.sh]: DOne

**Comment by bhapas** [comment]
test integrations

File: .buildkite/scripts/integration_tests.sh
Comment (mrodm) [inline .buildkite/scripts/integration_tests.sh:27]:
```suggestion
FALSE_POSITIVES_TARGET="test-check-packages-false-positives"
```

File: scripts/test-check-false-positives.sh
Comment (mrodm) [inline scripts/test-check-false-positives.sh:28]: Currently, with these changes even if the new step runs successfully, there are some JUnit annotations because there were some expected tests failing.

To avoid showing those Junit annotations in Buildkite, I think one option that could we do here is to delete the `build/test-results/*.xml` files in the cleanup function if "$r == 1" . WDYT ? @bhapas 

Junit annotations with errors

Also, in case there were no errors, some echo message could be added there to explain why it is failing.


```suggestion
    if [ $r == 1 ]; then
      rm build/test-results/*.xml
      exit 0
    elif [ $r == 0 ]; then
      echo "Expected to fail tests, but there was none failing"
      exit 1
    fi
```

cc @jsoriano

File: scripts/test-check-false-positives.sh
Comment (mrodm) [inline scripts/test-check-false-positives.sh:26]: Just wondering if this could be leveraged to check specifically which tests are failing.

It could happen that some other test is failing and not the expected one. Or another scenario it would be that more tests are failing than expected.

I see that the resulting xml is like this 

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites>
  <testsuite name="system" tests="1" failures="1">
    <!--test suite for system tests-->
    <testcase name="system test: pagination" classname="httpjson_false_positive_asserts.generic" time="28.841040253">
      <failure>observed hit count 4 did not match expected hit count 2</failure>
    </testcase>
  </testsuite>
</testsuites>
```

Maybe we could add another file next to the folder `test/packages/false_positives/httpjson_false_positive_asserts`, something like:
`test/packages/false_positives/httpjson_false_positive_asserts-expected_errors.xml` or similar

and compare somehow those two xmls.

There would be differences because I see there is a field `time="XX.XX"`. Could it be checked at least the failures lines ? Or a line by line comparison with regexes ? 

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites>
  <testsuite name="system" tests="1" failures="1">
    <!--test suite for system tests-->
    <testcase name="system test: pagination" classname="httpjson_false_positive_asserts.generic" time="*">
      <failure>observed hit count 4 did not match expected hit count 2</failure>
    </testcase>
  </testsuite>
</testsuites>
```

Just testing with the contents of the xml:
- a new file could be added that contains the expected errors:
    - it can contain regexes and double quotes must be escaped.
  ```
  <testcase name=\"system test: pagination\" classname=\"httpjson_false_positive_asserts.generic\" time=\".*\"> * <failure>observed hit count 4 did not match expected hit count 2</failure>
  ```
- Then in this `scripts/test-check-packages-false-positives` it could be added the comparison line by line to check for these errors

Comment (mrodm) [reply scripts/test-check-false-positives.sh]: In this commit, I pushed an approach to deal with this
https://github.com/elastic/elastic-package/pull/1333/commits/51e33f4d8e93fc18b7eae030911167fb0f150f55

Let me know what you think @jsoriano @bhapas

Comment (mrodm) [reply scripts/test-check-false-positives.sh]: That changeset is supposed to fail the build (https://buildkite.com/elastic/elastic-package/builds/1142), I will do a couple of tests more
- same number of errors, but different one
- expected errors match with the actual errors

Comment (jsoriano) [reply scripts/test-check-false-positives.sh]: LGTM. Maybe at some point we should migrate these scripts to go, they are starting to have some complexity.

Comment (mrodm) [reply scripts/test-check-false-positives.sh]: First build with the expected output
- Error: There are unexpected errors in httpjson_false_positive_asserts

Comment (bhapas) [reply scripts/test-check-false-positives.sh]: Test Expected Error Scenarios:

- [ ]  Number of expected and actual errors are not equal
- [ ]  Number of expected and actual errors are equal but different contents in the actual errors than expected
- [ ]  The Expected and actual errors are same

**Review by mrodm — APPROVED** [review]
:+1: Thanks for adding this negative tests @bhapas!

File: docs/howto/system_testing.md
Comment (jsoriano) [inline docs/howto/system_testing.md:586]: Thanks for adding docs for this too!

**Review by jsoriano — APPROVED** [review]

**Pull request merged**
